clippy_lints/src/casts/ptr_cast_constness.rs

```diff
@@ -26,20 +27,87 @@ pub(super) fn check<'tcx>(
-        let constness = match *to_mutbl {
-            Mutability::Not => "const",
-            Mutability::Mut => "mut",
-        };
+        let mut app = Applicability::MachineApplicable;
+        if let ExprKind::Call(func, []) = cast_expr.kind
+            && let ExprKind::Path(QPath::Resolved(None, path)) = func.kind
+            && let Some(defid) = path.res.opt_def_id()
+            && let Some(prefix) = std_or_core(cx)
+            && let sugg = format!("{}", Sugg::hir_with_applicability(cx, cast_expr, "_", &mut app))
+            && let Some((_, after_lt)) = sugg.split_once("::<")
+        {
+            if cx.tcx.is_diagnostic_item(sym::ptr_null, defid) {
+                span_lint_and_sugg(
+                    cx,
+                    PTR_CAST_CONSTNESS,
+                    expr.span,
+                    "`as` casting to make a const null pointer into a mutable null pointer",
+                    "use `null_mut()` directly instead",
+                    format!("{prefix}::ptr::null_mut::<{after_lt}"),
+                    app,
+                );
+                return;
+            }
+            if cx.tcx.is_diagnostic_item(sym::ptr_null_mut, defid) {
+                span_lint_and_sugg(
+                    cx,
+                    PTR_CAST_CONSTNESS,
+                    expr.span,
+                    "`as` casting to make a mutable null pointer into an immutable null pointer",
+                    "use `null()` directly instead",
+                    format!("{prefix}::ptr::null::<{after_lt}"),
+                    app,
+                );
+                return;
+            }
+        }
 
-        span_lint_and_sugg(
-            cx,
-            PTR_CAST_CONSTNESS,
-            expr.span,
-            "`as` casting between raw pointers while changing only its constness",
-            format!("try `pointer::cast_{constness}`, a safer alternative"),
-            format!("{}.cast_{constness}()", sugg.maybe_par()),
-            Applicability::MachineApplicable,
-        );
+        if msrv.meets(msrvs::POINTER_CAST_CONSTNESS) {
+            let sugg = Sugg::hir(cx, cast_expr, "_");
+            let constness = match *to_mutbl {
+                Mutability::Not => "const",
+                Mutability::Mut => "mut",
+            };
+
+            span_lint_and_sugg(
+                cx,
+                PTR_CAST_CONSTNESS,
+                expr.span,
+                "`as` casting between raw pointers while changing only its constness",
+                format!("try `pointer::cast_{constness}`, a safer alternative"),
+                format!("{}.cast_{constness}()", sugg.maybe_par()),
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
+pub(super) fn check_null_ptr_cast_method(cx: &LateContext<'_>, expr: &Expr<'_>) {
+    if let ExprKind::MethodCall(method, cast_expr, [], _) = expr.kind
+        && matches!(method.ident.as_str(), "cast_mut" | "cast_const")
```

Comment:
Can you move this check further down to be after the hir checks and store the method ident in a variable somewhere so the interner only needs to be locked once? Currently there are three calls to Symbol::as_str

Reply:
Also this `matches` and the ptr/ptr_mut diagnostic item checks can be merged together, would simplify the code a bit
```suggestion
        && let mutable = match (cx.tcx.get_diagnostic_name(defid), method.ident.as_str()) {
            (Some(sym::ptr_null), "cast_mut") => true,
            (Some(sym::ptr_null_mut), "cast_const") => false,
            _ => return,
        };
```

Reply:
I moved the `let` statements into the `if` (which simplifies further the previous `let Some(…) else { return; };`). Indeed, using `get_diagnostic_name()` makes it easier than `is_diagnostic_name()` here.